src/main.rs

```diff
@@ -179,6 +187,117 @@ unsafe fn search256<W: Write>(bytes: &[u8], mut output: &mut W) -> Result<(), st
+fn search128<W: Write>(bytes: &[u8], mut output: &mut W) -> Result<(), std::io::Error> {
+    use core::arch::aarch64::*;
+
+    let ptr = bytes.as_ptr();
+    let mut last_printed = bytes.len();
+    let mut index = last_printed - 1;
+
+    if index >= 64 {
+        // ARMv8 loads do not have alignment *requirements*, but there can be performance penalties
+        // (e.g. seems to be about 2% slowdown on Cortex-A72 with a 500MB file) so let's align.
+        // Search unaligned bytes via slow method so subsequent haystack reads are always aligned.
+        let align_offset = unsafe { ptr.offset(index as isize).align_offset(16) };
+        let aligned_index = index as usize + align_offset - 16;
+
+        // eprintln!("Unoptimized search from {} to {}", aligned_index, last_printed);
+        slow_search_and_print(
+            bytes,
+            aligned_index,
+            last_printed,
+            &mut last_printed,
+            &mut output,
+        )?;
+        index = aligned_index;
+        drop(aligned_index);
+
+        let pattern128 = unsafe { vdupq_n_u8(SEARCH) };
+        while index >= 64 {
+            let window_end_offset = index;
+            unsafe {
+                index -= 16;
+                let window = ptr.add(index);
+                let search128 = vld1q_u8(window);
+                let result128_0 = vceqq_u8(search128, pattern128);
+
+                index -= 16;
+                let window = ptr.add(index);
+                let search128 = vld1q_u8(window);
+                let result128_1 = vceqq_u8(search128, pattern128);
+
+                index -= 16;
+                let window = ptr.add(index);
+                let search128 = vld1q_u8(window);
+                let result128_2 = vceqq_u8(search128, pattern128);
+
+                index -= 16;
+                let window = ptr.add(index);
+                let search128 = vld1q_u8(window);
+                let result128_3 = vceqq_u8(search128, pattern128);
+
+                // Bulk movemask as described in
+                // https://branchfree.org/2019/04/01/fitting-my-head-through-the-arm-holes-or-two-sequences-to-substitute-for-the-missing-pmovmskb-instruction-on-arm-neon/
+                let mut matches = {
+                    let bit_mask: uint8x16_t = std::mem::transmute([
+                        0x01u8, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80, 0x01, 0x02, 0x4, 0x8, 0x10,
+                        0x20, 0x40, 0x80,
+                    ]);
+                    let t0 = vandq_u8(result128_3, bit_mask);
+                    let t1 = vandq_u8(result128_2, bit_mask);
+                    let t2 = vandq_u8(result128_1, bit_mask);
+                    let t3 = vandq_u8(result128_0, bit_mask);
+                    let sum0 = vpaddq_u8(t0, t1);
+                    let sum1 = vpaddq_u8(t2, t3);
+                    let sum0 = vpaddq_u8(sum0, sum1);
+                    let sum0 = vpaddq_u8(sum0, sum0);
```

Comment:
It's annoying that this right here is why it's ARM64-only. `vpadd_u8` (no `q`) is available on ARM but it's going to take a lot more operations to get the result that way and probably not worth it.

Reply:
Who would ever want to chew through huge log files and whatnot at SIMD speeds on old 32-bit junk anyway? :D